internal/archive/credentials.go

```diff
@@ -0,0 +1,304 @@
+package archive
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+)
+
+// This file defines findCredentials() function for searching repository
+// credentials in Apt configuration, see
+// https://manpages.debian.org/testing/apt/apt_auth.conf.5.en.html.
+
+// credentials contains matched non-empty Username and Password.
+// Username is left empty if the search is unsuccessful.
+type credentials struct {
+	Username string
+	Password string
+}
+
+// Empty checks whether c represents unsuccessful search.
+func (c credentials) Empty() bool {
+	return c.Username == ""
+}
+
+// credentialsQuery contains parsed input URL data used for search.
+type credentialsQuery struct {
+	scheme     string
+	host       string
+	port       string
+	path       string
+	needScheme bool
+}
+
+// queryFor parses repoUrl into credentialsQuery and fills provided credentials with
+// username and password if they are specified in repoUrl.
+func queryFor(repoUrl string, creds *credentials) (*credentialsQuery, error) {
```

Comment:
This API needs love. The function name gives no hints of what it's querying for, half of its result is done via a parameter, half via a result, with a result that looks like a question ("query").

Compare it with:
```
func parseCredentialsQuery(repoURL string) (*credentialsQuery, error) { ... }
```
Username and Password should be part of the query result itself, after all it was indeed provided as the query data, I guess?

Also, trivial nitpick: let's please keep acronyms uppercased, unless they must start the word in which case it's all lowercased.

Reply:
The reason the `credentials` structure is passed as a pointer argument is because it's defined as a return value (not a pointer) in `findCredentialsInDir`. It's passed here only for the rare case where the archive URL has username and password in it (`https://foo:bar@example.com/...`), in which case, those credentials are returned early from `findCredentialsInDir`. This case won't happen in Chisel but the behavior mirrors Apt's parser.

The name is indeed bad. I'm not sure the proposed one is better. The input of this function is an archive URL and the return value is the input of the parser (except for the case where credentials are specified in URL) which I call query now but I'm not so sure about it.

What do you mean by _query data_? The username and password is what we're looking for in the netrc files.

Reply:
I did the following changes:
- Changed name to `parseRepoURL`
- Dropped the pointer return argument
- Added `credentials` return value

The reason `credentials` is not embedded in `credentialsQuery` is that when username is specified in `repoURL`, `credentialsQuery` does not need to be allocated.

Are these changes OK?